usaspending_api/database_scripts/tableview_generator/tableview_sql_generator.py

```diff
@@ -0,0 +1,417 @@
+    total = len(create_indexes)
+    print_debug("There are {} index creations".format(total))
+    indexes_and_msg = []
+    for n, index in enumerate(create_indexes):
+        if n % 10 == 0 and n > 0:
+            console = TEMPLATE["sql_print_output"].format("{} indexes created, {} remaining".format(n, total - n))
+            indexes_and_msg.append(console)
+        indexes_and_msg.append(index)
+
+    return indexes_and_msg, rename_old_indexes, rename_new_indexes
+
+
+def make_modification_sql(name):
+    global CLUSTERING_INDEX
+    sql_strings = []
+    if CLUSTERING_INDEX:
+        print_debug("*** This tableview will be clustered on {} ***".format(CLUSTERING_INDEX))
+        sql_strings.append(TEMPLATE["cluster"].format(name, CLUSTERING_INDEX))
+    sql_strings.append(TEMPLATE["analyze"].format(name))
+    sql_strings.append(TEMPLATE["grant_select"].format(name, "readonly"))
+    return sql_strings
+
+
+def make_rename_sql(name, old_indexes, new_indexes):
+    temp_name = name + "_temp"
+    archive_name = name + "_old"
+    sql_strings = []
+    sql_strings.append(TEMPLATE["rename"].format("IF EXISTS ", name, archive_name))
+    sql_strings += old_indexes
+    sql_strings.append("")
+    sql_strings.append(TEMPLATE["rename"].format("", temp_name, name))
+    sql_strings += new_indexes
+    return sql_strings
+
+
+def create_all_sql_strings(sql_json):
+    """ Desired ordering of steps for final SQL:
+        1. Drop existing "_temp" and "_old" tableviews
+        2. Create new tableview
+        3. Create indexes for new tableview
+        4. (optional) Cluster tableview on index
+        5. analyze verbose <tableview>
+        6. Rename existing tableview, append with _old
+        7. Rename all existing tableview indexes to avoid name collisions
+        8. Rename new tableview
+        9. Rename new tableview indexes
+       10. Clean up
+    """
+    final_sql_strings = []
+
+    name = sql_json["name"]
+    temp_name = name + "_temp"
+
+    create_indexes, rename_old_indexes, rename_new_indexes = make_indexes_sql(sql_json, temp_name)
+
+    final_sql_strings.extend(make_drops(name))
+    final_sql_strings.append("")
+    final_sql_strings.extend(make_create(name, sql_json))
+
+    final_sql_strings.append("")
+    final_sql_strings += create_indexes
+    final_sql_strings.append("")
+    final_sql_strings.extend(make_rename_sql(name, rename_old_indexes, rename_new_indexes))
+    final_sql_strings.append("")
+    final_sql_strings.extend(make_modification_sql(name))
+    final_sql_strings.append("")
+    final_sql_strings.extend(make_clean_up(sql_json))
+
+    return final_sql_strings
+
+
+def write_sql_file(str_list, filename):
+    fname = filename + ".sql"
+
+    print_debug("Creating file: {}".format(fname))
+    with open(fname, "w") as f:
+        fstring = "\n".join(str_list)
+        f.write(fstring)
+        f.write("\n")
+
+
+def create_componentized_files(sql_json):
```

Comment:
One of my issues with the approach to generating all of these from a JSON definition is that it trends towards a code-generation pattern (shudder) rather than just defining your code and version-controlling changes to it.

That is, why not actually just save the `.sql` files in git version control. When they need to change, change them. That way we can track what changed. They don't even need to be changed by hand: run the generator to change them, just store the versioned result.

... Perhaps the constraints of having to do side-by-side recreate + redeploy led us to this to avoid naming collisions and conduct swaps+renames?